DeeperFunc/StringParser.swift

```diff
@@ -0,0 +1,163 @@
+//
+//  StringParser.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 31/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+typealias StringParser<S: PatternState> = ([String]) -> (rest: [String], match: [RoutePattern<Any, S>])?
+
+extension Array {
+    func flatMapFirst<T>(where transform: (Element) throws -> T?) rethrows -> T? {
+        for element in self {
+            if let transformed = try transform(element) { return transformed }
+        }
+        return nil
+    }
+}
+
+func parseComponents<S>(with parsers: [StringParser<S>]) -> StringParser<S> {
+    return { components in
+        var components = components
+        var pathPatterns = [RoutePattern<Any, S>]()
+        while !components.isEmpty {
```

Comment:
can we use the same approach as with routes, using composition and recursion instead of loop?